org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/internal/Breakpoint.java

```diff
@@ -0,0 +1,199 @@
+import io.reactivex.disposables.Disposable;
+
+public class Breakpoint implements IBreakpoint {
+    private VirtualMachine vm = null;
+    private IEventHub eventHub = null;
+    private String className = null;
+    private int lineNumber = 0;
+    private int hitCount = 0;
+
+    Breakpoint(VirtualMachine vm, IEventHub eventHub, String className, int lineNumber) {
+        this(vm, eventHub, className, lineNumber, 0);
+    }
+
+    Breakpoint(VirtualMachine vm, IEventHub eventHub, String className, int lineNumber, int hitCount) {
+        this.vm = vm;
+        this.eventHub = eventHub;
+        this.className = className;
+        this.lineNumber = lineNumber;
+        this.hitCount = hitCount;
+    }
+
+    // IDebugResource
+    private List<EventRequest> requests = new ArrayList<EventRequest>();
+    private List<Disposable> subscriptions = new ArrayList<Disposable>();
+
+    @Override
+    public List<EventRequest> requests() {
+        return requests;
+    }
+
+    @Override
+    public List<Disposable> subscriptions() {
+        return subscriptions;
+    }
+
+    // AutoCloseable
+    @Override
+    public void close() throws Exception {
+        vm.eventRequestManager().deleteEventRequests(requests());
+        subscriptions().forEach(subscription -> {
+            subscription.dispose();
+        });
+    }
+
+    // IBreakpoint
+    @Override
+    public String className() {
+        return className;
+    }
+
+    @Override
+    public int lineNumber() {
+        return lineNumber;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (!(obj instanceof IBreakpoint)) {
+            return super.equals(obj);
+        }
+
+        IBreakpoint breakpoint = (IBreakpoint)obj;
+        return this.className().equals(breakpoint.className()) && this.lineNumber() == breakpoint.lineNumber();
+    }
+
+    @Override
+    public int hitCount() {
+        return hitCount;
+    }
+
+    @Override
+    public void setHitCount(int hitCount) {
+        this.hitCount = hitCount;
+
+        Observable.fromIterable(this.requests())
+        .filter(request -> request instanceof BreakpointRequest)
+        .subscribe(request -> {
+            request.addCountFilter(hitCount);
+            request.enable();
+        });
+    }
+
+    @Override
+    public CompletableFuture<IBreakpoint> install() {
+        // It's possible that different class loaders create new class with the same name.
+        // Here to listen to future class prepare events to handle such case.
+        ClassPrepareRequest classPrepareRequest = vm.eventRequestManager().createClassPrepareRequest();
+        classPrepareRequest.addClassFilter(className);
+        classPrepareRequest.enable();
+        requests.add(classPrepareRequest);
+
+        // Local types also needs to be handled
+        ClassPrepareRequest localClassPrepareRequest = vm.eventRequestManager().createClassPrepareRequest();
+        localClassPrepareRequest.addClassFilter(className + "$*");
```

Comment:
className may be like xxx$1. Here we should parse the first level enclosing type name and add classFilter for enclosing$*